code/go/0chain.net/chaincore/chain/entity.go

```diff
@@ -618,43 +618,49 @@ func (c *Chain) setupInitialState(initStates *state.InitStates) util.MerklePatri
 	return pmt
 }
 
-func (c *Chain) addInitialStakes(mpt *util.MerklePatriciaTrie, stakes []state.InitStake) error {
-	var edbDelegatePools []event.DelegatePool
+func (c *Chain) addInitialStakes(stakes []state.InitStake, balances cstate.StateContextI) error {
+	var edbDelegatePools []*event.DelegatePool
+	knownStakePoolKeys := map[string]bool{}
 	for _, v := range stakes {
-		key := v.ProviderType + ":stakepool:" + v.ProviderID
+		providerType := spenum.ToProviderType(v.ProviderType)
+		stakePoolKey := stakepool.StakePoolKey(providerType, v.ProviderID)
 		sp := stakepool.StakePool{}
-		if err := mpt.GetNodeValue(util.Path(encryption.Hash(key)), &sp); err != nil {
-			if err != util.ErrValueNotPresent {
-				continue
+		if knownStakePoolKeys[stakePoolKey] {
```

Comment:
We don't need to create a map to check if the stakePool exist, checking the MPT by calling `sp.Get()` is enough, we won't re-enter this function later after we refactor it to only run genesis block once. So let's keep the changes simple, and avoid exposing the `stakepool.StakePoolKey()` function as well

Reply:
Got it. One of the reasons to add the map was to add up the tokens for stakes with same clientID and and provider. It would also reduce calls to mpt for each entry in stakepool. But if that's fine, I'll  remove it.

Reply:
We don't need to add up the tokens for stakes, we would only initialize stakes **once** on genesis block, which is round 1. It's a deterministic result, and we shall never change the staked value on genesis block later again. So please remove the `add up token` code and only do existence checking.

Reply:
okay, so removing code to support adding up stakes for cases as below:
```
  - provider_id: x
    provider_type: blobber
    client_id: y
    tokens: 50000000000

  - provider_id: x
    provider_type: blobber
    client_id: y
    tokens: 10000000000
```
resulting in tokens = 60000000000